src/Common/StringSearcher.h

```diff
@@ -826,66 +826,44 @@ using UTF8CaseInsensitiveStringSearcher = StringSearcher<false, false>;
-  * It makes sense to use only with short haystacks when cheap initialization is required.
-  * There is no option for case-insensitive search for UTF-8 strings.
-  * It is required that strings are zero-terminated.
-  */
-
-struct LibCASCIICaseSensitiveStringSearcher : public StringSearcherBase
+/// Use only with short haystacks where cheap initialization is required.
+template <bool CaseInsensitive>
+struct StdLibASCIIStringSearcher : public StringSearcherBase
 {
-    const char * const needle;
-
-    template <typename CharT>
-    requires (sizeof(CharT) == 1)
-    LibCASCIICaseSensitiveStringSearcher(const CharT * const needle_, const size_t /* needle_size */)
-        : needle(reinterpret_cast<const char *>(needle_)) {}
+    const char * const needle_start;
+    const char * const needle_end;
 
     template <typename CharT>
     requires (sizeof(CharT) == 1)
-    const CharT * search(const CharT * haystack, const CharT * const haystack_end) const
-    {
-        const auto * res = strstr(reinterpret_cast<const char *>(haystack), reinterpret_cast<const char *>(needle));
-        if (!res)
-            return haystack_end;
-        return reinterpret_cast<const CharT *>(res);
-    }
+    StdLibASCIIStringSearcher(const CharT * const needle_start_, const size_t needle_size_)
+        : needle_start{reinterpret_cast<const char *>(needle_start_)}
+        , needle_end{reinterpret_cast<const char *>(needle_start) + needle_size_}
+    {}
 
     template <typename CharT>
     requires (sizeof(CharT) == 1)
-    const CharT * search(const CharT * haystack, const size_t haystack_size) const
+    const CharT * search(const CharT * haystack_start, const CharT * const haystack_end) const
     {
-        return search(haystack, haystack + haystack_size);
-    }
-};
-
-struct LibCASCIICaseInsensitiveStringSearcher : public StringSearcherBase
-{
-    const char * const needle;
-
-    template <typename CharT>
-    requires (sizeof(CharT) == 1)
-    LibCASCIICaseInsensitiveStringSearcher(const CharT * const needle_, const size_t /* needle_size */)
-        : needle(reinterpret_cast<const char *>(needle_)) {}
-
-    template <typename CharT>
-    requires (sizeof(CharT) == 1)
-    const CharT * search(const CharT * haystack, const CharT * const haystack_end) const
-    {
-        const auto * res = strcasestr(reinterpret_cast<const char *>(haystack), reinterpret_cast<const char *>(needle));
-        if (!res)
-            return haystack_end;
-        return reinterpret_cast<const CharT *>(res);
+        if constexpr (CaseInsensitive)
+        {
+            return std::search(
+                haystack_start, haystack_end, needle_start, needle_end,
+                [](char c1, char c2) {return std::toupper(c1) == std::toupper(c2);});
+        }
+        else
+        {
+            return std::search(
+                haystack_start, haystack_end, needle_start, needle_end,
+                [](char c1, char c2) {return c1 == c2;});
```

Comment:
Just had a complete wtf moment when I realized after multiple hours of debugging how the failures in test "00233_position_function_family" were caused. If the haystack or needle contains multi-byte/unicode characters and we want to search case-sensitively, then a plain `std::search(haystack_start, haystack_end, needle_start, needle_end);` finds no matches. It does however if we provide a comparator to `std::sort`. Oddly, the comparator is the one which I would have expected to be used anyways ... Unfortunately, my debugger didn't step into libcpp so I couldn't really find out what was going on.